components/dashboard/src/hooks/use-user-loader.ts

```diff
@@ -0,0 +1,58 @@
+/**
+ * Copyright (c) 2022 Gitpod GmbH. All rights reserved.
+ * Licensed under the GNU Affero General Public License (AGPL).
+ * See License.AGPL.txt in the project root for license information.
+ */
+
+import { useState, useContext } from "react";
+import { User } from "@gitpod/gitpod-protocol";
+import { UserContext } from "../user-context";
+import { getGitpodService } from "../service/service";
+import { ErrorCodes } from "@gitpod/gitpod-protocol/lib/messaging/error";
+import { trackLocation } from "../Analytics";
+import { refreshSearchData } from "../components/RepositoryFinder";
+import { useQuery } from "@tanstack/react-query";
+import { noPersistence } from "../data/setup";
+
+export const useUserLoader = () => {
+    const { user, setUser } = useContext(UserContext);
+    const [isSetupRequired, setSetupRequired] = useState(false);
+
+    // For now, we're using the user context to store the user, but letting react-query handle the loading
+    // In the future, we should remove the user context and use react-query to access the user
+    const { isLoading } = useQuery({
+        queryKey: noPersistence(["current-user"]),
```

Comment:
I think this should actually be the `user.id || "not-logged-in-user"`. Or what happens to the cache if I logout and login with a different user? What mechanism ensures it gets invalidated? And if there is such a mechanism: Does it make sense to rely on it, or should it be built-in here? :thinking:

Reply:
Couple things:

* We don't have the user yet here, so we don't know their `id` to be able to include it in the query key.
* The `noPersistence()` wrapper around the `queryKey` ensures we don't reload this key from the query cache on page load.
* I'll check to see if we're wiping the query cache on logout, we should be doing that if we're not, good call (will do in a followup if we're not yet).

I think it would be an improvement if we could use the logged in user's id as part of the cache key for where we store the query cache in indexed db. This would help reduce any risk of accidentally loading a different user's query cache if 2 shared the same browser. It's a bit of a chicken/egg scenario since we don't have access to the logged in user's id until after we've loaded them though. I think it could be possible to adjust some of the way the app bootstraps though so we can handle loading the user, and then setup a new react query QueryCache w/ the user's id available. I need to play around w/ some of that a bit though to understand if it's feasible, but I think it would be a nice improvement.